consensus/misc/eip1559/eip1559.go

```diff
@@ -141,47 +63,65 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header, time uint64)
 	}
 	elasticity := config.ElasticityMultiplier()
 	denominator := config.BaseFeeChangeDenominator(time)
+	var minBaseFee *uint64
+
+	// OPStack addition: from Holocene onwards, decode
+	// denominator, elasticity and minBaseFee from
+	// the extra data using optimism-specific rules.
 	if config.IsHolocene(parent.Time) {
-		denominator, elasticity = DecodeHoloceneExtraData(parent.Extra)
-		if denominator == 0 {
-			// this shouldn't happen as the ExtraData should have been validated prior
-			panic("invalid eip-1559 params in extradata")
-		}
+		denominator, elasticity, minBaseFee = DecodeOptimismExtraData(config, parent.Time, parent.Extra)
 	}
+
 	parentGasTarget := parent.GasLimit / elasticity
-	// If the parent gasUsed is the same as the target, the baseFee remains unchanged.
-	if parent.GasUsed == parentGasTarget {
-		return new(big.Int).Set(parent.BaseFee)
-	}
 
-	var (
-		num   = new(big.Int)
-		denom = new(big.Int)
-	)
+	// OPStack addition: use a closure to calculate the base fee using
+	// the upstream code.
+	baseFee := func() *big.Int {
```

Comment:
The closure unfortunately still produces a large diff (if not ignoring white spaces) because all code is now indented...